File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/RefListParameter.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class RefListParameter {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class RefListParameterCleaner implements Runnable {

        MemorySegment segment;
        RefListParameterCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.RefListParameter_destroy(this.segment);
        }
    }

    RefListParameter() {}
    RefListParameter(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/RefListParameter.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface RefListParameterLib: Library {
    fun RefListParameter_destroy(handle: Pointer)
}

class RefListParameter internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class RefListParameterCleaner(val handle: Pointer, val lib: RefListParameterLib) : Runnable {
        override fun run() {
            lib.RefListParameter_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, RefListParameter.RefListParameterCleaner(handle, RefListParameter.lib));
    }

    companion object {
        internal val libClass: Class<RefListParameterLib> = RefListParameterLib::class.java
        internal val lib: RefListParameterLib = Native.load("diplomat_feature_tests", libClass)
    }

}